src/scratchOrgInfoGenerator.ts

```diff
@@ -0,0 +1,298 @@
+      hubOrg.getConnection().getAuthInfoFields().clientId ?? defaultConnectedAppInfo.clientId;
+  }
+
+  if (!nonamespace && sfdxProject?.get('namespace')) {
+    scratchOrgInfoPayload.namespace = sfdxProject.get('namespace') as string;
+  }
+
+  // we already have the info, and want to get rid of configApi, so this doesn't use that
+  scratchOrgInfoPayload.connectedAppCallbackUrl = `http://localhost:${await WebOAuthServer.determineOauthPort()}/OauthRedirect`;
+  return scratchOrgInfoPayload;
+};
+
+/**
+ * Returns a valid signup json
+ *
+ * @param definitionjson
+ * @param definitionfile
+ * @param connectedAppConsumerKey
+ * @param durationdays
+ * @param nonamespace
+ * @param noancestors
+ * @param orgConfig
+ * @returns scratchOrgInfoPayload: ScratchOrgInfoPayload;
+            ignoreAncestorIds: boolean;
+            warnings: string[]; 
+ */
+export const getScratchOrgInfoPayload = async (options: {
+  definitionjson: string;
+  definitionfile: string;
+  connectedAppConsumerKey: string;
+  durationDays: number;
+  nonamespace: boolean;
+  noancestors: boolean;
+  orgConfig: Record<string, unknown>;
+}): Promise<{
+  scratchOrgInfoPayload: ScratchOrgInfoPayload;
+  ignoreAncestorIds: boolean;
+  warnings: string[];
+}> => {
+  const warnings: string[] = [];
+  // Varargs input overrides definitionjson (-j option; hidden/deprecated)
+  const definitionJson = options.definitionjson ? JSON.parse(options.definitionjson) : {};
+  const orgConfigInput = { ...definitionJson, ...(options.orgConfig ?? {}) };
+
+  let scratchOrgInfoPayload = orgConfigInput;
+
+  // the -f option
+  if (options.definitionfile) {
+    try {
+      const fileData = await fs.readFile(options.definitionfile, 'utf8');
+      const defFileContents = parseJson(fileData) as Record<string, unknown>;
+      // definitionjson and varargs override file input
+      scratchOrgInfoPayload = { ...defFileContents, ...orgConfigInput };
+    } catch (err) {
+      const error = err as Error;
+      if (error.name === 'SyntaxError') {
+        throw new SfdxError(`An error occurred parsing ${options.definitionfile}`);
+      }
+    }
+  }
+
+  // scratchOrgInfoPayload must be heads down camelcase.
+  const upperCaseKey = sfdc.findUpperCaseKeys(scratchOrgInfoPayload);
+  if (upperCaseKey) {
+    throw new SfdxError('InvalidJsonCasing', upperCaseKey);
+  }
+
+  // Now run the fully resolved user input against the validator
+  Object.keys(scratchOrgInfoPayload).forEach((key) => {
+    optionsValidator(key, scratchOrgInfoPayload);
+  });
+
+  if (options.connectedAppConsumerKey) {
+    scratchOrgInfoPayload.connectedAppConsumerKey = options.connectedAppConsumerKey;
+  }
+
+  scratchOrgInfoPayload.durationDays = options.durationDays;
+
+  // Throw warnings for deprecated scratch org features.
+  const scratchOrgFeatureDeprecation = new ScratchOrgFeatureDeprecation();
+  scratchOrgFeatureDeprecation.getFeatureWarnings(scratchOrgInfoPayload.features).forEach((warning) => {
```

Comment:
I think this belongs after the `if (typeof scratchOrgInfoPayload.features === 'string') {` code (which ensures the features are an array).

That would simplify what has to happen in `getFeatureWarnings` (it could always expect an array)